# Crazy Lawn Mower
- Based on the example narrated by Justin Saddlemyer in Probability Puzzles Webinar
- Robotic Lawn mower is placed at the center of the lawn, but it tends to move randomly without a pattern.
- The possible movements by the robotic lawn mower are :
	- Standard moving distance = 1 m
	- Horizontal = forward (+1) or backward (-1)
	- Vertical  = forward (+1) or backward (-1)

# Problem Statement
- The owner of the lawn mower places the lawn mower in the centre of a lawn 40m by 40m
- The owner leaves the house for a given period of time (in minutes)
- By the time the owner returns, what is the **probability that crazy lawn mower stays inside the lawn ?** 

In [29]:
# away_time of owner from mower
away_time = 700

In [30]:
# function to simulate mower movement
def mower_sim(away_time: int)-> int:
    """
    simulate the random movement coordinates of mower in a given time
    
    Args:
    ----
    away_time - time in minutes, the owner is away from the robotic lawn mower
    
    Returns:
    -------
    x_pos, y_pos - lists of x and y coordinates of lawn mower movement
    """
    # import numpy
    import numpy as np
    
    # mower position
    mower_x = 0
    mower_y = 0

    # mower movement coordinates list
    x_pos = []
    y_pos = []
    move_range = range(-1, 2)

    # mower movement for owner's away_time
    for move in range(away_time):
        mower_x += np.random.choice(move_range)
        mower_y += np.random.choice(move_range)
        x_pos.append(mower_x)
        y_pos.append(mower_y)
    
    return x_pos, y_pos


# owner away_time (in mins) from the mower
x_pos, y_pos = mower_sim(away_time=away_time)
print(x_pos[:10], y_pos[:10])

[-1, -2, -3, -4, -5, -4, -3, -4, -3, -4] [1, 2, 1, 0, 1, 1, 1, 2, 2, 2]


In [34]:
# function to visualize mower movement
def mower_tracker(away_time):
    """
    calls mower_sim function and visualizes the tracks of mower using simulated coordinates
    
    Args:
    -----
    away_time - time in minutes, the owner is away from the robotic lawn mower 
    
    Returns:
    -------
    visualization of the mower movement  
    """
    # import plotly
    import plotly.express as px
    
    # call mower_sim
    x_pos, y_pos = mower_sim(away_time=away_time)
    
    # line chart
    fig = px.line(x=x_pos, y=y_pos, width=700, height=700, title='Crazy Lawn Mower moves!')
    
    # lawn boundaries
    fig.add_shape(x0=-40, x1=40, y0=-40, y1=40, line_color='red', fillcolor='green', opacity=0.2)
    
    # total lawn area and beyond
    fig.update_xaxes(range=[-50, 50])
    fig.update_yaxes(range=[-50, 50])
    
    fig.show()
    
    return None


# visualize mower tracks for away_time (in minutes)
mower_tracker(away_time=away_time)
    

In [35]:
# function to check if mower stays inside lawn or not
def mower_stays(away_time: int)-> int:
    """
    calls mower_sim function and determines if mower is inside the lawn or not
    
    Returns:
    -------
    boolean, True if mower inside lawn else False
    """
    # call mower_sim
    x_pos, y_pos = mower_sim(away_time=away_time)

    # if mower stays inside lawn
    if max(x_pos) <= 50 and min(x_pos) >= -50 and max(y_pos) <= 50 and min(y_pos) >= -50 :
        return True
    else:
        return False

# test run mower_stays
mower_stays(away_time=away_time)

True

In [33]:
# run 10000 simulation iterations of mower movement
sim_runs = []
for run in range(10000):
    sim_runs.append(mower_stays(away_time=away_time))
    
print(f"Probability of mower to stay inside lawn : {np.mean(sim_runs): .4}")

Probability of mower to stay inside lawn :  0.9284


# Conclusion
- Given a limited area(40m by 40m) and a Crazy robotic lawn mower and after 10000 simulations for an away_time = 700 minutes, **there is a high probability (0.9284) it STAYS inside lawn**